File: python-sdks/instrumentations/respan-instrumentation-exa/src/respan_instrumentation_exa/_serialization.py
"""Safe serialization for Exa request and response objects."""

from __future__ import annotations

import dataclasses
import json
from collections.abc import Mapping, Sequence
from datetime import date, datetime
from enum import Enum
from typing import Any

_MAX_DEPTH = 16
_SENSITIVE_MARKERS = frozenset(
    {
        "apikey",
        "apitoken",
        "authorization",
        "authtoken",
        "bearertoken",
        "cookie",
        "credential",
        "password",
        "privatekey",
        "refreshtoken",
        "secret",
        "sessiontoken",
        "accesstoken",
        "xapikey",
    }
)


def is_sensitive_key(key: object) -> bool:
    compact = "".join(char for char in str(key).lower() if char.isalnum())
    return compact == "token" or any(marker in compact for marker in _SENSITIVE_MARKERS)


def type_name(value: Any) -> str:
    value_type = type(value)
    return f"{value_type.__module__}.{value_type.__qualname__}"


def to_jsonable(value: Any, *, depth: int = 0, seen: set[int] | None = None) -> Any:
    """Convert Exa/Pydantic values to JSON data without leaking credentials."""

    if value is None or isinstance(value, (str, int, float, bool)):
        return value
    if depth > _MAX_DEPTH:
        return {"type": type_name(value), "truncated": True}
    if isinstance(value, bytes):
        return value.decode("utf-8", errors="replace")
    if isinstance(value, (date, datetime)):
        return value.isoformat()
    if isinstance(value, Enum):
        return to_jsonable(value.value, depth=depth + 1, seen=seen)

    if seen is None:
        seen = set()
    value_id = id(value)
    if value_id in seen:
        return {"type": type_name(value), "recursive": True}

    if isinstance(value, Mapping):
        seen.add(value_id)
        try:
            return {
                str(key): (
                    "<redacted>"
                    if is_sensitive_key(key)
                    else to_jsonable(item, depth=depth + 1, seen=seen)
                )
                for key, item in value.items()
                if not callable(item)
            }
        finally:
            seen.discard(value_id)

    if isinstance(value, Sequence) and not isinstance(value, (str, bytes, bytearray)):
        seen.add(value_id)
        try:
            return [to_jsonable(item, depth=depth + 1, seen=seen) for item in value]
        finally:
            seen.discard(value_id)

    model_dump = getattr(value, "model_dump", None)
    if callable(model_dump):
        try:
            return to_jsonable(
                model_dump(mode="json", exclude_none=True),
                depth=depth + 1,
                seen=seen,
            )
        except TypeError:
            return to_jsonable(model_dump(), depth=depth + 1, seen=seen)

    if dataclasses.is_dataclass(value) and not isinstance(value, type):
        return to_jsonable(dataclasses.asdict(value), depth=depth + 1, seen=seen)

    as_dict = getattr(value, "dict", None)
    if callable(as_dict):
        try:
            return to_jsonable(as_dict(), depth=depth + 1, seen=seen)
        except Exception:  # noqa: BLE001, S110 - third-party serializers are best effort.
            pass

    if hasattr(value, "__dict__"):
        public = {
            key: item
            for key, item in vars(value).items()
            if not key.startswith("_") and not callable(item)
        }
        if public:
            return to_jsonable(public, depth=depth + 1, seen=seen)

    return {"type": type_name(value), "value": str(value)}


def json_dumps(value: Any) -> str:
    return json.dumps(
        to_jsonable(value),
        ensure_ascii=False,
        separators=(",", ":"),
        sort_keys=True,
    )


def value_at(value: Any, name: str, default: Any = None) -> Any:
    if isinstance(value, Mapping):
        return value.get(name, default)
    return getattr(value, name, default)

File: javascript-sdks/instrumentations/respan-instrumentation-exa/src/_serialization.ts
const MAX_DEPTH = 16;
const SENSITIVE_MARKERS = new Set([
  "apikey",
  "apitoken",
  "authorization",
  "authtoken",
  "bearertoken",
  "cookie",
  "credential",
  "password",
  "privatekey",
  "refreshtoken",
  "secret",
  "sessiontoken",
  "accesstoken",
  "xapikey",
]);

export function isSensitiveKey(key: PropertyKey): boolean {
  const compact = String(key).toLowerCase().replace(/[^a-z0-9]/g, "");
  if (compact === "token") return true;
  return [...SENSITIVE_MARKERS].some((marker) => compact.includes(marker));
}

export function toSerializable(
  value: unknown,
  depth = 0,
  seen: WeakSet<object> = new WeakSet(),
): unknown {
  if (
    value === null ||
    value === undefined ||
    typeof value === "string" ||
    typeof value === "number" ||
    typeof value === "boolean"
  ) {
    return value ?? null;
  }
  if (typeof value === "bigint") return value.toString();
  if (typeof value === "symbol" || typeof value === "function") {
    return { type: typeof value };
  }
  if (depth > MAX_DEPTH) {
    return { type: typeName(value), truncated: true };
  }
  if (value instanceof Date) return value.toISOString();
  if (value instanceof Error) {
    return { name: value.name, message: value.message };
  }
  if (typeof value !== "object") return String(value);
  if (seen.has(value)) return { type: typeName(value), recursive: true };
  seen.add(value);
  try {
    if (Array.isArray(value)) {
      return value.map((item) => toSerializable(item, depth + 1, seen));
    }
    if (value instanceof Map) {
      return Object.fromEntries(
        [...value.entries()].map(([key, item]) => [
          String(key),
          isSensitiveKey(key)
            ? "<redacted>"
            : toSerializable(item, depth + 1, seen),
        ]),
      );
    }
    if (value instanceof Set) {
      return [...value].map((item) => toSerializable(item, depth + 1, seen));
    }
    const record = value as Record<string, unknown>;
    if (typeof record.toJSON === "function") {
      try {
        return toSerializable(record.toJSON(), depth + 1, seen);
      } catch {
        // Fall through to enumerable fields.
      }
    }
    return Object.fromEntries(
      Object.entries(record)
        .filter(([, item]) => typeof item !== "function")
        .map(([key, item]) => [
          key,
          isSensitiveKey(key)
            ? "<redacted>"
            : toSerializable(item, depth + 1, seen),
        ]),
    );
  } finally {
    seen.delete(value);
  }
}

export function safeJson(value: unknown): string {
  return JSON.stringify(toSerializable(value));
}

export function typeName(value: unknown): string {
  if (value && typeof value === "object") {
    return (value as { constructor?: { name?: string } }).constructor?.name ?? "object";
  }
  return typeof value;
}

export function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value && typeof value === "object" && !Array.isArray(value));
}

export function valueAt(value: unknown, key: string): unknown {
  return isRecord(value) ? value[key] : undefined;
}
